숫자 야구 게임 [SangBeom-Hahn]

SangBeom-Hahn opened the pull request:
> model 구현, 테스트 완료 / controller 구현 중

SangBeom-Hahn committed:
docs(README): list requirements and feature specifications

SangBeom-Hahn committed:
feat(Computer): 상대방 컴퓨터 객체 구현

SangBeom-Hahn committed:
test(Computer): 상대방 컴퓨터 객체 테스트

SangBeom-Hahn committed:
refactor(Computer): 상대방 컴퓨터 랜덤 수 문자열이었던 것 List로 리팩토링

- stream()을 사용하기 위해 String이었던 랜덤 수를 List로 변환
- 테스트를 위해 private을 public으로 변환

SangBeom-Hahn committed:
feat(User): 사용자 객체 구현

SangBeom-Hahn committed:
feat(User): 숫자를 대입하는 setNumber() 기능 추가

SangBeom-Hahn committed:
test(User): 사용자 객체 테스트

SangBeom-Hahn committed:
feat(BaseballController): 사용자가 랜덤 수를 맞추면 재시작(1) / 종료(2) 기능 구현

SangBeom-Hahn committed:
test(BaseballController): 사용자가 랜덤 수를 맞추면 재시작(1) / 종료(2) 기능 테스트

sjiwon commented on docs/README.md:
```diff
@@ -0,0 +1,39 @@
+### 요구사항 내용
+1. 컴퓨터
+   - 1 ~ 9까지 서로 다른 임의의 수 3개를 선택한다.
+   - 사용자가 입력한 숫자에 대한 결과를 출력한다.
```
일상적인 관점에서는 컴퓨터가 결과를 출력하지만 이 게임상에서는 컴퓨터 & 사용자를 Player로 보고 결과를 도출하는 것은 다른 컴포넌트에게 위임하는게 좋지 않을까요?

sjiwon commented on src/main/java/baseball/controller/BaseballController.java:
```diff
@@ -0,0 +1,51 @@
+    public void terminateGame(User user) {
+        if (isTerminate()) {
+            user.terminate();
+            System.out.println("완전 정료");
```
이런 출력은 요구사항에 없는데 임의로 넣으신건가요

sjiwon commented on src/main/java/baseball/controller/BaseballController.java:
```diff
@@ -0,0 +1,51 @@
+            user.restart();
+        }
+    }
+    public boolean isTerminate() {
```
개행

sjiwon commented on src/main/java/baseball/controller/BaseballController.java:
```diff
@@ -0,0 +1,51 @@
+        validateNotStringRestartStatus(restartStatus);
+
+
+        if(restartStatus.equals("2") ){
```
2줄을 개행한 이유가 있나요

그리고 똑같은 검증을 2번하고 있네요

sjiwon commented on src/main/java/baseball/controller/BaseballController.java:
```diff
@@ -0,0 +1,51 @@
+            throw new IllegalArgumentException("재시작은 1, 완전 종료는 2인 정수로 소수를 입력할 수 없습니다.");
+        }
+    }
+
```
`validateRangeRestartStatus`를 가장 먼저 검증하고 있는데
- Int가 아닌 값
- 숫자가 아닌 값
- 중간에 Space Separator 존재

이 경우는 어떻게 대응하실건가요?

그리고 restart에 필요한 검증을 크게보면
1. 사용자가 숫자를 입력했냐
2. 그 숫자가 1, 2중에 하나인가

이렇게 나올거같은데 위의 3가지 검증은 포커스가 어긋난거 같은데 어떻게 생각하시나요?
- 정수냐 소수냐 -> 따로 정수 판별에 대한 boolean method로 판별하면 될거같은데

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+public class Computer {
+    public List<Integer> randomNumber;
+
+    public Computer() {
+        saveRandomNumberWithGameStart();
+    }
```
save를 통해서 randomNumber에 대한 init을 한다는거는 알겠는데
위의 코드보다는

```suggestion
public Computer() {
    this.randomNumber = generate~~();
}
```

와 같이 수정하는게 더 가독성이 좋을거 같은데 어떻게 생각하시나요?

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+        return randomNumber;
+    }
+
+    public void saveRandomNumberWithGameStart() {
+        randomNumber = new ArrayList<>();
+        Integer digit;
+
+        while (checkLengthSmallThanThree()) {
+            digit = getRandomDigit();
+            if (!hasDuplicateDigitInRandomNumber(digit)) {
+                randomNumber.add(digit);
+            }
+        }
+    }
```
테스트를 위해서 열어두신거 같은데 굳이 얘를 열어둘 필요가 있을까요?
 new Computer()로 생성에 대해서 테스트 하면 같이 테스트 되는거 아닌가요?

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    public boolean checkLengthSmallThanThree() {
+        if (randomNumber.size() < 3) {
+            return true;
+        }
+        return false;
+    }
```
1줄로 줄일 수 있을거 같네요

그리고 `3`이란 숫자가 물론 이 미션의 명세를 보면 뭘 의미하는지 알겠지만 코드 레벨에서 무슨 숫자인지 파악이 가능할까요?

- 이렇게 코드 레벨에서 의미를 나타내지 않는거를 매직넘버라고 합니다

추가적으로 미션에서는 숫자는 3개만 받도록 제한을 뒀지만 만약에 5개로 늘리면 이 메소드명까지 영향을 받는거 아닌가요?

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+        return Randoms.pickNumberInRange(1, 9);
+    }
+
+    public boolean hasDuplicateDigitInRandomNumber(Integer digit) {
+        if (randomNumber.contains(digit)) {
+            return true;
+        }
+        return false;
+    }
```
마찬가지로 1줄로 줄일 수 있네요

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+        saveRandomNumberWithGameStart();
+    }
+
+    public List<Integer> getRandomNumber() {
+        return randomNumber;
+    }
```
일반적으로 getter/setter/equals와 같은 메소드들은 클래스 가장 하단에 위치하는게 관례입니다
- 외부에서 변경할 이유가 없다면 mutable -> immutable

sjiwon commented on src/main/java/baseball/model/User.java:
```diff
@@ -0,0 +1,34 @@
+        this.number = number;
+    }
+
+    public void validateNumber(final List<Integer> number) {
+        if (number.size() > 3 || number.size() < 3) {
+            throw new IllegalArgumentException("입력 숫자는 3자리입니다.");
+        }
+
+        if (!number.stream().allMatch(digit -> digit >= 1 && digit <= 9)) {
+            throw new IllegalArgumentException("입력 숫자는 3자리입니다.");
+        }
+    }
```
매직넘버는 위에서 말했듯이 따로 관리해서 가독성 좋게 표현해주세요

그리고 if 인자로 stream을 한줄로 배치하셨는데 가독성이 너무 안좋지 않을까요?
차라리 메소드로 빼서 가독성을 좋게 하는게 나을거같네요

Exception Message만 보면 동일한 메시지를 넘기는데 2가지 검증의 차이가 드러나게 표현해주세요

sjiwon commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+public class Computer {
+    public List<Integer> randomNumber;
```
이 필드는 왜 public으로 열어둔거죠?

sjiwon commented on src/test/java/baseball/model/ComputerTest.java:
```diff
@@ -0,0 +1,68 @@
+
+        // then
+        assertThat(randomNumber.stream().allMatch(digit -> digit >= 111 && digit <= 999));
+        System.out.println(randomNumber);
```
테스트 코드에서 sout으로 출력을 하는 이유가 있나요

sjiwon reviewed (commented):
필요한 부분
- `Optimize Import` 생활화
- 매직넘버 및 코드 가독성에 대한 고민을 가지는 시간
- 파일 끝 EOF 신경써서 처리
- 기본적인 접근 제한자 개념

sjiwon commented on src/test/java/baseball/model/ComputerTest.java:
```diff
@@ -0,0 +1,68 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+
+class ComputerTest {
```
각각의 테스트가 무엇을 위한 테스트인지는 알겠는데 private scope로 두어야 할 메소드들도 테스트를 위해서 public으로 열어둔 것처럼 보이네요

클래스의 모든 메소드를 테스트하기 위해서 public으로 열어두기 보다 private scope로 두어야할 메소드는 private scope로 두고 해당 메소드를 활용하는 의미있는 public method를 테스트하면 되지 않을까요?

sjiwon replied:
각각의 테스트가 무엇을 위한 테스트인지는 알겠는데 private scope로 두어야 할 메소드들도 테스트를 위해서 public으로 열어둔 것처럼 보이네요

클래스의 모든 메소드를 테스트하기 위해서 public으로 열어두기 보다 private scope로 두어야할 메소드는 private scope로 두고 해당 메소드를 활용하는 의미있는 public method를 테스트하면 되지 않을까요?

> https://shoulditestprivatemethods.com/

sjiwon commented on src/test/java/baseball/model/UserTest.java:
```diff
@@ -0,0 +1,55 @@
+        User user = new User();
+
+        // when
+        ArrayList<Integer> number = new ArrayList<>(Arrays.asList(1, 2));
```
타입을 List로 하지않고 ArrayList로 둔 이유가 있나요?

추가적으로 `new ArrayList<>(Arrays.asList(1, 2))`로 구현한 이유가 있나요? 그냥 `Arrays.asList(1, 2)`로 해도 되지않아요?
- 그리고 `Arrays.asList와 List.of`의 차이도 뭔지 알고계시나요?

sjiwon commented on src/test/java/baseball/model/UserTest.java:
```diff
@@ -0,0 +1,55 @@
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("RestartStatus 테스트")
+    void UserTest() {
+        // given
+        User user = new User();
+
+        // when
+        user.terminate();
+
+        // then
+        assertThat(user.restartStatus).isEqualTo(User.RestartStatus.RESTART);
+    }
```
이 테스트 깨지는데요
기본적으로 PR을 생성할때는 구현한 로직에 대해서 로컬에서 모든 테스트가 통과하는지 확인하고 올려주세요

SangBeom-Hahn committed:
refactor(BaseballController): 사용자가 랜덤 수를 맞추면 재시작(1) / 종류(2) 기능 리팩토링

- 종료 신호 사용자 입력 메서드 분리
- BaseballControllerTest 코드 terminateTest 메서드에서 console 입력 부분 String 변수 초기화로 대체

SangBeom-Hahn committed:
feat(BaseballController): 사용자 숫자 입력 시 결과 반환(스트라이크, 볼, 낫싱) 기능 구현

SangBeom-Hahn committed:
test(BaseballController): 사용자 숫자 입력 시 결과 반환(스트라이크, 볼, 낫싱) 기능 테스트

- 테스트를 위해 컴퓨터의 생성자에 save 메서드 주석처리

SangBeom-Hahn committed:
feat(BaseballController): 게임 시작 기능 구현

SangBeom-Hahn committed:
feat(BaseballController): 사용자 3자리 숫자 입력 기능 구현

SangBeom-Hahn committed:
test(BaseballController): 사용자 3자리 숫자 입력 기능 테스트

SangBeom-Hahn committed:
feat(BaseballController): 사용자 3자리 숫자 입력에 따른 결과 반환 기능 구현

SangBeom-Hahn committed:
test(BaseballController): 사용자 3자리 숫자 입력에 따른 결과 반환 기능 테스트

- controller의 playGame 메서드에 주석은 resultTest 출력을 보기 위한 것으로 이렇게 해도 되는지 질문하고 싶습니다.

SangBeom-Hahn committed:
feat(BaseballController) : 게임 종료 기능 구현

SangBeom-Hahn committed:
test(BaseballController) : 게임 종료 기능 테스트

SangBeom-Hahn committed:
refactor(BaseballController) : 게임 시나리오를 위한 run 메서드 리팩토링

- user의 RestartStatus에 재시작 여부를 입력하던 terminateGame 메서드를 삭제
- gameOver 메서드에서 terminateGame 메서드 제거
- 게임 재시작시 컴퓨터의 랜덤수가 동일한 문제 해결

SangBeom-Hahn committed:
refactor(OutputView) : 하드코딩 했던 것 view 분리 및 상수로 리팩토링

SangBeom-Hahn committed:
refactor(InputView) : 하드코딩 했던 것 view 분리 및 상수로 리팩토링

SangBeom-Hahn committed:
refactor(BaseballController) : 하드코딩 했던 것 상수로 리팩토링

SangBeom-Hahn committed:
refactor(User) : 하드코딩 했던 것 상수로 리팩토링

SangBeom-Hahn committed:
refactor(BaseballController) : User, Computer 지역변수에서 전역 변수로 리팩토링

SangBeom-Hahn committed:
refactor(BaseballController) : strikeCount, ballCount 지역변수에서 전역 변수로 리팩토링

SangBeom-Hahn committed:
refactor(BaseballController) : strike, ball 카운트 메서드 stream으로 리팩토링

SangBeom-Hahn committed:
fix(OutputView) : strike, ball의 갯수가 0일 때 0도 출력되는 버그 수정

SangBeom-Hahn committed:
refactor : 접근 지정자 리팩토링

SangBeom-Hahn commented:
일단은 제 방식대로 리팩토링했고 커멘트에 대한 답변과 리팩토링을 할 예정입니다.

SangBeom-Hahn committed:
refactor : 검증 추가 리팩토링
